src/source/Utils/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,221 @@
+        case BACKOFF_TERMINATED:
+            DLOGI("Cannot execute exponentialBackoffBlockingWait. Current status is BACKOFF_TERMINATED");
+            retStatus = STATUS_EXPONENTIAL_INVALID_STATE;
+            break;
+        default:
+            DLOGI("Cannot execute exponentialBackoffBlockingWait. Unexpected state [%d]", pExponentialBackoffState->status);
+            retStatus = STATUS_EXPONENTIAL_INVALID_STATE;
+    }
+
+    CleanUp:
+        LEAVES();
+        return retStatus;
+}
+
+STATUS exponentialBackoffBlockingWait(PExponentialBackoffState pRetryState) {
+    ENTERS();
+    PExponentialBackoffConfig pRetryConfig = NULL;
+    UINT64 currentTime = 0, waitTime = 0, jitter = 0;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pRetryState != NULL, STATUS_NULL_ARG);
+    CHK(pRetryState->pExponentialBackoffConfig != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(validateAndUpdateExponentialBackoffStatus(pRetryState));
+
+    pRetryConfig = pRetryState->pExponentialBackoffConfig;
+
+    // If retries is exhausted, return error to the application
+    if (pRetryConfig->maxRetryCount != INFINITE_RETRY_COUNT_SENTINEL) {
+        CHK(!(pRetryConfig->maxRetryCount == pRetryState->currentRetryCount),
+            STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED);
+    }
+
+    // check the last retry time. If the difference between last retry and current time is greater than
+    // minTimeToResetRetryState, then reset the retry state
+    //
+    // Proceed if this is not the first retry
+    currentTime = GETTIME();
+    if (pRetryState->currentRetryCount != 0) {
+        if (currentTime - pRetryState->lastRetryWaitTime > pRetryConfig->minTimeToResetRetryState) {
+            CHK_STATUS(resetRetryState(pRetryState));
+            pRetryState->status = BACKOFF_IN_PROGRESS;
+        }
+    }
+
+    // Bound the exponential curve to maxWaitTime. Once we reach
+    // maxWaitTime, then we always wait for maxWaitTime time
+    // till the state is reset.
+    if (pRetryState->lastRetryWaitTime == pRetryConfig->maxWaitTime) {
+        waitTime = pRetryState->lastRetryWaitTime;
+    } else {
+        waitTime = calculateWaitTime(pRetryState, pRetryConfig);
+        if (waitTime > pRetryConfig->maxWaitTime) {
+            waitTime = pRetryConfig->maxWaitTime;
+        }
+    }
+
+    jitter = getRandomJitter(pRetryConfig->jitterFactor);
+    THREAD_SLEEP(waitTime + jitter);
+
+    pRetryState->lastRetryWaitTime = currentTime + waitTime + jitter;
+    pRetryState->currentRetryCount++;
+    DLOGI("Number of retries done [%"PRIu32"]. Last retry wait time [%"PRIu64"]",
+          pRetryState->currentRetryCount, pRetryState->lastRetryWaitTime);
+
+CleanUp:
+    if (retStatus == STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED) {
+        DLOGI("Exhausted exponential retries");
+        resetRetryState(pRetryState);
+    }
+    LEAVES();
+    return retStatus;
+}
+
+VOID freeExponentialBackoffState(PExponentialBackoffState* ppExponentialBackoffState) {
```

Comment:
Suggest making this signature with return type STATUS. If we ever want to expand this to include status codes, we would not have to make change this. Considering this is a PUBLIC_API, we need to provide flexibility to avoid unnecessary backward compat issues.

Reply:
I had that earlier but it was not used. So I removed it. But your point is valid. I can add it back.

Reply:
Yep and in addition once it goes into PIC, `STATUS` return type for `free*` APIs is consistent with other public APIs.  To be really consistent to match existing naming conventions I would call this `exponentialBackoffStateFree`. Similar naming for other APIs to match what is already in PIC.